kernel/src/hil/nfc.rs

```diff
@@ -0,0 +1,105 @@
+use crate::returncode::ReturnCode;
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+/// An enum to keep track of the NFC field status
+pub enum NfcFieldState {
+    /// Initial value that indicates no NFCT field events.
+    None,
+    /// The NFCT FIELDLOST event has been set.
+    On,
+    /// The NFCT FIELDDETECTED event has been set.
+    Off,
+    /// Both NFCT field events have been set - ambiguous state.
+    Unknown,
+}
+
+/// Controls an NFC tag main functionalities
+pub trait NfcTag<'a> {
+    /// Set the client instance that will handle callbacks
+    fn set_client(&self, client: &'a dyn Client<'a>);
+
+    /// Enable NFC sense field mode, and subscribe to the relevant interrupts.
+    /// Also set up the default configurations of how frame delay should be
+    /// dealt with (e.g. the maximum delay before a timeout).
+    /// This function should never fail.
+    fn enable(&self);
+
+    /// Enable tag emulation by triggering the necessary task.
+    /// This function should never fail.
+    fn activate(&self);
```

Comment:
What are the "necessary task"s here?

"should" is too weak, if the function _cannot_ fail, we can say so. However, it seems possible that some NFC implementations may exist but fail to activate for various reasons. This probably needs a `ReturnCode`